Android refactoring + rtc stats

vona-ben opened the pull request:
Note: A dev branch should be created so we can merge this after review

Complete refactoring of Android code. Using a similar approach of OPENTOK network library JS.

Added `NetworkQualityTest` class is a utility for testing network quality and gathering statistics during video streaming sessions using the OpenTok platform. It allows monitoring various network and media statistics, such as video bitrate, audio quality, round-trip time, and more.

The NetworkQualityTest will send regular callbacks which can be handled using NetworkQualityTestCallbackListener.

The `MainActivity`class is a sample that shows how to use this class and how to handle the callback.

<img width="504" alt="Screenshot 2023-08-30 at 5 40 43 PM" src="https://github.com/opentok/opentok-network-test/assets/108658664/7d300aff-d10b-488b-9019-229a1e2fd532">

Commit:
WIP update android stats

Commit:
Remove credentiels

Commit:
Clean-up

Commit:
Refactoring

Commit:
Fix subscriber stas

josemrecio committed:
Typo in annotation

josemrecio committed:
Fix buiding settings

josemrecio committed:
Better estimation of availableOutgoingBitrate

vona-ben committed:
Merge pull request #1 from josemrecio/android-rtc-stats-update

Better estimation of availableOutgoingBitrate

Commit:
[WIP] Refactoring

Commit:
WIP refactoring

Commit:
WIP refactoring

Commit:
WIP refactoring

Commit:
WIP refactoring

Commit:
WIP refactoring

Commit:
Refactoring

Commit:
WIP refactoring

Commit:
WIP refactoring

Commit:
WIP refactoring

Commit:
Remove lombodk

Commit:
clean up

Commit:
Add a method to stop the test

Commit:
Clean-up

juliobecerragomez commented on Android-Sample/app/src/main/java/com/opentok/qualitystats/sample/NetworkQualityTest.java:
```diff
@@ -0,0 +1,617 @@
+                }
+                jitter = rtcStatObject.optDouble(JITTER, -1);
+            } else if (CANDIDATE_PAIR.equals(statType)) {
+                boolean isNominated = rtcStatObject.optBoolean(IS_NOMINATED, false);
```
I think several candidates may be nominated. Why not using the `selectedCandidatePairId` from the transport stats?

https://developer.mozilla.org/en-US/docs/Web/API/RTCIceCandidatePairStats/selected

goncalocostamendes commented on Android-Sample/app/src/main/java/com/opentok/qualitystats/sample/MainActivity.java:
```diff
@@ -2,407 +2,205 @@
-    private AlertDialog dialog;
+public class MainActivity extends Activity {
+    static final String LOGTAG = "quality-stats-demo";
+    private static final String SESSION_ID = "YOUR_SESSION_ID";
```
For the most distracted users, let's add some documentation on the README about this. By not being empty anymore the app will no longer alert the user and will try to connect for the duration of the timeout.

goncalocostamendes commented on Android-Sample/README.md:
````diff
@@ -16,217 +20,108 @@ The sample app only subscribes to the test stream. It does not subscribe to othe
-    mHandler.postDelayed(statsRunnable, TEST_DURATION * 1000);
-}
-```
+    To stop the test, you can call stopTest. If the test is halted before 5 seconds, an error will be triggered.
````
nit: change "you can call stopTest" to "you can call NetworkQualityTest.stopTest()"

goncalocostamendes commented on Android-Sample/README.md:
```diff
@@ -1,7 +1,11 @@
 
-This sample shows how to use this OpenTok Android SDK to determine the appropriate audio and video
+# NetworkQualityTest
+
```
nit: Link to NetworkQualityTest README

goncalocostamendes commented on Android-Sample/app/README.md:
```diff
@@ -0,0 +1,94 @@
+
+6. **Clean Up**
+
+   To ensure proper resource release, you should handle cleanup when needed. In your `onPause`
```
If I am not mistaken, onPause() is triggered when app goes to the background. Video is off, but audio is still on. Having that in mind, don't know if there is the need to stop the test (of course, knowing without video, the stats will go up). Your call

goncalocostamendes commented on Android-Sample/app/src/main/AndroidManifest.xml:
```diff
@@ -9,6 +8,8 @@
     <uses-permission android:name="android.permission.MODIFY_AUDIO_SETTINGS" />
     <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
```
WRITE_EXTERNAL_STORAGE permission does not have to be requested on the java side as well? On java side only see camera, internet and audio.

vona-ben replied:
Seems is not require but I'm not expert on that :D

goncalocostamendes commented on Android-Sample/app/src/main/java/com/opentok/qualitystats/sample/MainActivity.java:
```diff
@@ -2,407 +2,205 @@
+                        Log.d(LOGTAG, "Audio Packet Lost Ratio  " + stats.getAudioPacketLostRatio() * 100 + "%");
+                        Log.d(LOGTAG, "Video Packet Lost Ratio  " + stats.getVideoPacketLostRatio() * 100 + "%");
+                        Log.d(LOGTAG, "Jitter: " + stats.getJitter());
+                        Log.d(LOGTAG, "Quality Limitation Reason: " + stats.getQualityLimitationReason());
```
Maybe we can change the previous logs with `Log.d(LOGTAG, sentVideoInfo);`. In the future if we have to change something, we only have to change it in one place

goncalocostamendes commented on Android-Sample/app/src/main/java/com/opentok/qualitystats/sample/MainActivity.java:
```diff
@@ -2,407 +2,205 @@
+                    }
+                });
+        // Start the quality test
+        networkQualityTest.startTest();
```
Should we only start the test when we start publishing? What if the user does not provide the permissions, or takes a lot of time to do so, and the test finishes without even having published? (Based on what I understood, if not explicitly stopped the test only takes 30s)

vona-ben replied:
I moved permission to the MainActivity and only run the test now if the user accepted the permissions.

goncalocostamendes commented on Android-Sample/README.md:
````diff
@@ -16,217 +20,108 @@ The sample app only subscribes to the test stream. It does not subscribe to othe
+    ```java
+    new NetworkQualityTestCallbackListener() {
+        @Override
+        public void onQualityTestResults(String recommendedSetting) {
````
Can the `recommendedSetting` be `do not use video` or something similar? If the minimum conditions are not met?

vona-ben replied:
There is a threshold where the video is not recommended, added it to the docs and comments

goncalocostamendes reviewed (commented):
Overall looks great.
I left some comments. The majority are **nit**. 
Tomorrow I will test the app to do a final validation

jeffswartz commented on Android-Sample/README.md:
```diff
@@ -16,217 +20,108 @@ The sample app only subscribes to the test stream. It does not subscribe to othe
+    
+    - `onQualityTestResults`: Receive the final test results and the recommended setting based on quality thresholds.
+    - `onQualityTestStatsUpdate`: Receive real-time statistics updates during the test.
+    - `onError`: Handle errors that occur during the test.
```
Please provide more details on the API:

* `onQualityTestResults()` -- What are the possible values of `recommendedSetting`, and what do they mean?
* `onQualityTestStatsUpdate()` -- What are the properties of `CallbackQualityStats`, and what do they mean?
* `onError()` -- What are the possible errors, and what do they mean?

vona-ben replied:
@jeffswartz I have update the docs: Missing the errors as it require more proper handle in the code

jeffswartz commented on Android-Sample/README.md:
```diff
@@ -16,217 +20,108 @@ The sample app only subscribes to the test stream. It does not subscribe to othe
-   * Voice-only -- Your bandwidth is too low for video.
+    You can configure the following parameters:
+    
+    - `sessionId`: The session ID of your OpenTok session.
```
Mention that this should be a [routed session](https://tokbox.com/developer/guides/create-session/#media-mode) (right?) And mention that this session should only be used for the test (not for communications).

jeffswartz commented on Android-Sample/README.md:
```diff
@@ -16,217 +20,108 @@ The sample app only subscribes to the test stream. It does not subscribe to othe
+    - `apiKey`: Your OpenTok API key.
+    - `token`: The token for connecting to the session.
+    - `resolution`: The camera capture resolution for the publisher. Defaults to `HIGH` if not specified.
+    - `testDurationSec`: The duration of the test in seconds. Defaults to `30 sec` if not specified.
```
What are possible values for `resolution` and `testDurationSec`.

jeffswartz commented on Android-Sample/app/README.md:
```diff
@@ -0,0 +1,94 @@
+# NetworkQualityTest
+
+The `NetworkQualityTest` class is a utility for testing network quality and gathering statistics
+during video streaming sessions using the OpenTok platform. It allows you to monitor various network
+and media statistics, such as video bitrate, audio quality, round-trip time, and more.
```
Add a link to the docs on the callback methods (see my comments on the main README).

Commit:
Fix permission issues + PR comments

Commit:
Fix permission issues

Commit:
Clean-up

vona-ben committed:
Update README.md

vona-ben committed:
Update README.md

vona-ben committed:
Update README.md

vona-ben committed:
Update README.md

Commit:
Before this change, the subscriber bitrate was calculated since the beginning, not in each interval.
This is because the reference data to compare with the latest one is not the previous, but the first.

vona-ben committed:
Merge pull request #3 from josemrecio/android-rtc-stats

Calculate subscriber bitrates for each interval, not since the beginning
